### PR TITLE
Add metrics to reversetunnel connections

### PR DESCRIPTION
**Pull request by dboslee**
These metrics will help to see the difference in latency between
connections over local tunnels, peer proxies, and direct dials.

The time to establish a connection is measured from the top of the function until the connection is returned.
The time to first use a connection is measured from when the connection is established to when the first read returns.

Implements part of https://github.com/gravitational/cloud/issues/1823

### CODE REVIEW

File: lib/reversetunnel/conn_metric.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+)
+
+// metricConn reports metrics for reversetunnel connections.
+type metricConn struct {
+	net.Conn
+	clock clockwork.Clock
+
+	// start is the time since the last state was reported.
+	start    time.Time
+	firstUse sync.Once
```
Comment (vitorenesduarte) [inline lib/reversetunnel/conn_metric.go:49]: Nit: we could maybe rename all instances of "first use" to "first read", or something else that better represents a connection `state` (the histogram label) like "established".

**Review by vitorenesduarte — APPROVED** [review]

**Review by rosstimothy — COMMENTED** [review]
This seems a bit

File: lib/reversetunnel/conn_metric.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+func (c *metricConn) duration() time.Duration {
+	now := c.clock.Now()
+	d := now.Sub(c.start)
+	c.start = now
+	return d
+}
+
+// addConn updates the connection and dial type. It also reports the time
+// it took to establish the connection.
+func (c *metricConn) addConn(conn net.Conn, dt dialType) {
```
[the remark this answers is not in the document]
Comment (espadolini) [reply lib/reversetunnel/conn_metric.go:71]: I might be ok with this design - I like being explicit with when we start the timer, and our dialing isn't always easily expressed with a single dialer function - but we should enforce that we can only add a connection once.

Perhaps we could just take the time before we begin the dial procedures and have a constructor for the wrapper that takes the `Conn`, whatever metadata we have and the "time in which we started dialing"?

[the remark this answers is not in the document]
Comment (dboslee) [reply lib/reversetunnel/conn_metric.go]: Updated here [5e6a132](https://github.com/gravitational/teleport/pull/14027/commits/5e6a13200fd6215747a97d97c58eaebd091fe815)

**Review by marcoandredinis — COMMENTED** [review]
It seems we are calculating the total time from dial until the end of the first read.
I would expect this to be two metrics:
- time to connect: checks for DNS and latency times (among other things)
- time to first read: checks for server responsiveness

I'm saying this because they usually represent different infrastructure, and having this context should be helpful

If we want to store the time from dial until the end of the first read I would name the metric.
`time_to_first_byte` seems to be a somewhat common expression to denote how long it takes from dial to the first read.

File: lib/reversetunnel/conn_metric.go (hunk 1 above)
Comment (marcoandredinis) [reply lib/reversetunnel/conn_metric.go:49]: +1 for `first read`
You can also change to another variation and add the `Once` suffix which seems common on the codebase

File: lib/reversetunnel/conn_metric.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+// newMetricConn returns a new metricConn
+func newMetricConn(conn net.Conn, dt dialType, start time.Time, clock clockwork.Clock) *metricConn {
+	return &metricConn{
+		Conn:     conn,
+		dialType: dt,
+		start:    start,
+		clock:    clock,
+	}
+}
```
Comment (vitorenesduarte) [inline lib/reversetunnel/conn_metric.go:61]:
```suggestion
func newMetricConn(conn net.Conn, dt dialType, start time.Time, clock clockwork.Clock) *metricConn {
	m := &metricConn{
		Conn:     conn,
		dialType: dt,
		start:    start,
		clock:    clock,
	}
	connLatency.WithLabelValues(string(c.dialType), "established").Observe(c.duration().Seconds())
	return m
}
```

Comment (dboslee) [reply lib/reversetunnel/conn_metric.go:61]: Yep totally missed that in the refactor :sweat_smile:

**Review by rosstimothy — APPROVED** [review]

**Review by marcoandredinis — APPROVED** [review]

**Pull request merged**